packages/web-media-box/src/pipelines/mse/mseManager.ts

```diff
@@ -0,0 +1,178 @@
+          }
+        });
+      }
+      const appendOperation: SourceBufferOperation = {
+        type: OperationType.APPEND,
+        operation
+      };
+      queue.push(appendOperation);
+      this.processQueue(bufferWrapper);
+    }
+  }
+
+  public removeData(mimeType: string, start: number, end: number = Infinity) {
+    const bufferWrapper = this.sourceBuffers.get(mimeType);
+    if (bufferWrapper) {
+      const buffer = bufferWrapper.buffer;
+      const queue = bufferWrapper.queue;
+      const operation = (): Promise<void> => {
+        return new Promise((resolve, reject) => {
+          const removeAllEventListeners = () => {
+            buffer.removeEventListener('updateend', updateEnd);
+            buffer.removeEventListener('error', onError);
+            buffer.removeEventListener('abort', onAbort);
+          };
+          const updateEnd = () => {
+            removeAllEventListeners();
+            this.logger.debug(`Remove complete to ${mimeType} SourceBuffer from ${start} to ${end}`);
+            resolve();
+          };
+          const onError = (event: Event) => {
+            removeAllEventListeners();
+            this.logger.warn(`Error ${event} removing data from SourceBuffer of type ${mimeType}`);
+            reject();
+          };
+          const onAbort = (event: Event) => {
+            removeAllEventListeners();
+            this.logger.debug(`Aborted ${event} removing data from SourceBuffer of type ${mimeType}`);
+            reject();
+          }
+          buffer.addEventListener('updateend', updateEnd);
+          buffer.addEventListener('error', onError);
+          buffer.addEventListener('abort', onAbort);
+          try {
+            buffer.remove(start, end);
+          } catch (error) {
+            this.logger.warn(`${error} cannot remove data for ${mimeType} to the SourceBuffer from ${start} to ${end}`);
+          }
+        });
+      }
+      const appendOperation: SourceBufferOperation = {
+        type: OperationType.REMOVE,
+        operation
+      };
+      queue.push(appendOperation);
+      this.processQueue(bufferWrapper);
+    }
+  }
+
+  public getDuration(): number {
+    return this.mediaSource.duration;
+  }
+
+  public setLiveSeekableRange(start: number, end: number) {
+    this.mediaSource.setLiveSeekableRange(start, end);
+  }
+
+  public clearLiveSeekableRange() {
+    this.mediaSource.clearLiveSeekableRange();
+  }
+
+  public getSource(): string {
+    return URL.createObjectURL(this.mediaSource);
+  }
+
+  public endOfStream(reason: EndOfStreamError) {
+    this.mediaSource.endOfStream(reason);
+  }
+
+  private async initMediaSource(): Promise<void> {
+    return new Promise((resolve) => {
+      const sourceOpen = () => {
+        this.mediaSource.removeEventListener('sourceopen', sourceOpen);
+        resolve();
+      };
+      this.mediaSource.addEventListener('sourceopen', sourceOpen);
+    });
+  }
+
+  private addSourceBuffer(mimeCodec: string) {
```

Comment:
we can mark async, and use await to automatically return Promise

Reply:
Do we need `addSourceBuffer` to be asynchronous? IIRC the only async process initializing the media source and source buffers is waiting on the `sourceopen` event.

Reply:
nvm I see what you mean. 😅